pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -66,10 +66,29 @@ This file implements string parsing and creation for NumPy datetime.
  *
  * Returns 0 on success, -1 on failure.
  */
+
+#define FORMAT_STARTSWITH(ch) \
+  if (exact) { \
+    if (!format_len || *format != ch) { \
+        goto parse_error; \
+    } \
+    ++format; \
+    --format_len; \
+  } else { \
+    if (format_len > 0) { \
+        if (*format != ch) { \
+            goto parse_error;  \
+        } \
+        ++format; \
+        --format_len; \
+    } \
+  } \
```

Comment:
super minor nit: maybe this cold be a bit easier to parse?
```c++
// Always error on character mismatch, or when format is exhausted in the exact case.
if (*format != ch || (exact && !format_len)) goto parse_error;
// Advance pointer if format is not exhausted
if (format_len) {
  ++format;
  --format_len;
}
```

Reply:
thanks for your suggestion! are you sure this is right though? if `exact=False` and we're at the end of the format string, won't `*format != ch` resolve to `True` and so go to the `parse_error`? At least, if I try this, then I get a few test failures, e.g.
```console
(.venv) megorelli@penguin:~/pandas-dev$ cat f.py 
import pandas as pd

pd.to_datetime(['2022-01-01'], format='%Y-%m', exact=False)

(.venv) megorelli@penguin:~/pandas-dev$ python f.py 
Traceback (most recent call last):
  File "/home/megorelli/pandas-dev/f.py", line 3, in <module>
    pd.to_datetime(['2022-01-01'], format='%Y-%m', exact=False)
  File "/home/megorelli/pandas-dev/pandas/core/tools/datetimes.py", line 1130, in to_datetime
    result = convert_listlike(argc, format)
  File "/home/megorelli/pandas-dev/pandas/core/tools/datetimes.py", line 441, in _convert_listlike_datetimes
    result, tz_parsed = objects_to_datetime64ns(
  File "/home/megorelli/pandas-dev/pandas/core/arrays/datetimes.py", line 2205, in objects_to_datetime64ns
    result, tz_parsed = tslib.array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 440, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 622, in pandas._libs.tslib.array_to_datetime
    raise ValueError(
ValueError: time data "2022-01-01" at position 0 doesn't match format "%Y-%m"
```

Reply:
Would changing to `if (format_len > 1)` solve that? This should prevent `format` going beyond the very last character.

Reply:
The format exhaustion is not tested outside of the macro, right?

Reply:
@MarcoGorelli , I have updated the condition so that `*format != ch` is conditioned on `format_len > 0`. Would that work now? I would still argue that changing this whole function to iterate over format is a  better approach, but I might be missing some use cases for which this is not the case...

Reply:
Yeah, that's cleaner, thanks!

Regarding iterating over `format` - not sure that'd be possible, because there are parts of the codebase which reach here just by specifying `require_iso8601` but without passing `format`. An option could be to do something like:
- if `format == ''` and `not exact` then use the current path,
- else if `format != ''`, then iterate over `format` until it's exhausted, then use the current codepath
but I think that would add even more complexity (if I've understood correctly)